CXEC-15848: UI for reopen/close request

Parthlakhani opened the pull request:
https://jira.tools.sap/browse/CXEC-15848

CMS component for reopen request button
CMS component for close request button
Reopen request dialog box with file upload and message
Changes in file upload to support custom button

Parthlakhani committed:
CXEC-15848: Add UI for reopen/close request
CMS component for reopen request button
CMS component for close request button
Reopen request dialog box with file upload and message

Parthlakhani committed:
remove console

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-close/customer-ticketing-close.component.spec.ts:
```diff
@@ -0,0 +1,25 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
Test if TicketStatus is OPEN, the button should be seen
Test if TicketStatus is not OPEN, the button should not be seen

Parthlakhani replied:
Yes, adding the unit tests.

bgambocjaviniar replied:
just to understand, you'll add unit test now or later? 

Because I recall the previous PR had no unit tests, but was said it will be done later https://github.com/SAP/spartacus/pull/16005/files#r931180804

Parthlakhani replied:
yes, that was the plan earlier. because most of the unit tests might be dependent on the core implementation.
I agreed to add this ones since it looked like it would be more on HTML side.

Parthlakhani replied:
Had a conversation with Christian, he thought the same as this unit tests looks more on HTML side.
But still there are dependencies on a service to mock the method. hence, we decided It would be better to add unit tests while adding core implementation.

Parthlakhani committed:
fix unit tests and prettier

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-close/customer-ticketing-close.component.spec.ts:
```diff
@@ -0,0 +1,26 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
To add:
Test if ticket status is CLOSED, the button should not render
Test if ticket status is OPEN, the button should render

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-reopen/customer-ticketing-reopen-dialog/customer-ticketing-reopen-dialog.component.html:
```diff
@@ -0,0 +1,66 @@
+        <cx-icon [type]="iconTypes.UPLOAD"></cx-icon>
+        <span class="cx-message-footer-text">Upload File</span>
+      </cx-file-upload>
+
```
Missing file upload note:
File Size Limit: 10MB
Maximum of one attachment per message

Parthlakhani replied:
yes missed that, adding.

Parthlakhani committed:
add default validations and validation checks

Parthlakhani committed:
add missing transaltion

Parthlakhani committed:
Add validations
add a method to file form validators to check the maximum file entries
add validations to customer ticketing file upload

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-reopen/customer-ticketing-reopen-dialog/customer-ticketing-reopen-dialog.component.spec.ts:
```diff
@@ -0,0 +1,34 @@
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
```
Test for characters lefts and data truncation in case of copy and paste
Test for form validity to force message is not empty

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-reopen/customer-ticketing-reopen.component.spec.ts:
```diff
@@ -0,0 +1,42 @@
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
```
Test for button not rendered when ticket status is not CLOSED
Maybe test open dialog
And test close dialog with X in top right, button cancel and submit

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-reopen/customer-ticketing-reopen.component.ts:
```diff
@@ -0,0 +1,48 @@
+    );
+
+    if (dialog) {
+      this.subscription.add(dialog.pipe(take(1)).subscribe());
```
What is this 1 ?  Maybe a constant will help

Parthlakhani replied:
it's when we only want 1st value coming from the stream. It's used like this everywhere in the project.

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing.service.ts:
```diff
@@ -1,8 +1,77 @@
+    ],
+  });
+
+  getTciketSubject(): Observable<string> {
```
```suggestion
  getTicketSubject(): Observable<string> {
```

khristian23 commented on feature-libs/customer-ticketing/components/customer-ticketing-reopen/customer-ticketing-reopen-dialog/customer-ticketing-reopen-dialog.component.ts:
```diff
@@ -35,6 +36,23 @@ export class CustomerTicketingReopenDialogComponent implements OnInit {
+  get getInputCharactersLimit(): number {
+    return (
+      this.customerTicketingConfig.customerTicketing?.inputCharactersLimit ||
+      2000
```
How abut a constant to replace 2000 like MAX_INPUT_CHARACTERS

Parthlakhani replied:
yes, better.

Parthlakhani committed:
Update feature-libs/customer-ticketing/components/customer-ticketing.service.ts

Co-authored-by: Christian Montoya <christian.montoya.b@gmail.com>

Parthlakhani committed:
Add file upload notes

Parthlakhani committed:
fix lint errors

bgambocjaviniar commented on feature-libs/customer-ticketing/core/customer-ticketing-constants.ts:
```diff
@@ -0,0 +1 @@
+export const MAX_INPUT_CHARACTERS = 2000;
```
will this be re-used in other places? If not, keep it within the reopen component as a property

bgambocjaviniar replied:
If you really want,
Can most likely re-use it for `feature-libs/customer-ticketing/core/config/default-customer-ticketing-config.ts`


maybe make the other set default value as constants here. All that being said, the array as well

Parthlakhani replied:
Yes, It will be re-used in messaging component.

bgambocjaviniar commented on feature-libs/customer-ticketing/core/config/default-customer-ticketing-config.ts:
```diff
@@ -1,5 +1,20 @@
 
 export const defaultCustomerTicketingConfig: CustomerTicketingConfig = {
-  customerTicketing: {},
+  customerTicketing: {
```
I know you used `asm` as reference most likely, but move this default to the root instead

ASM was part of the first code-splitting that we had done, but was 'outdated' in terms of conventions

bgambocjaviniar commented on feature-libs/customer-ticketing/root/model/augmented-core.model.ts:
```diff
@@ -0,0 +1,7 @@
+
+declare module '@spartacus/storefront' {
+  const enum LAUNCH_CALLER {
+    CUSTOMER_TICKETING = 'CUSTOMER_TICKETING',
```
👍

Parthlakhani committed:
refactor config and file upload
moved default config to root
added template reference to fix file upload button issue in import products

Parthlakhani committed:
Merge branch 'epic/customer-ticketing' into feature/CXEC-15848

Parthlakhani committed:
fix click issue on upload file

khristian23 commented on feature-libs/customer-ticketing/root/config/default-customer-ticketing-config.ts:
```diff
@@ -3,7 +3,7 @@ import { CustomerTicketingConfig } from './customer-ticketing-config';
 export const defaultCustomerTicketingConfig: CustomerTicketingConfig = {
   customerTicketing: {
     attachmentValidity: {
```
Validity or Validation?  or moreover: attachmentRestrictions

Parthlakhani replied:
yes that's better

Parthlakhani committed:
refactor object name

Parthlakhani committed:
fix prettier issue

Parthlakhani committed:
add todo

bgambocjaviniar commented on feature-libs/customer-ticketing/styles/_customer-ticketing-actions.scss:
```diff
@@ -0,0 +1,16 @@
+cx-customer-ticketing-close,
```
let's use placeholder selector to not put more technical debt in the future. 

Other features (that I did too 😓) should be using %selectorname 

https://sass-lang.com/documentation/style-rules/placeholder-selectors
https://sap.github.io/spartacus-docs/css-architecture/#placeholder-selectors

- Allows the use of @extend directive at a more root level than doing @extend .something
- we can skip things using our mechanism for skipping component styles

bgambocjaviniar replied:
You need to have things like this set `feature-libs/cart/base/styles/components/_index.scss`

Parthlakhani committed:
add placeholders selectors

Parthlakhani committed:
update css
